### PR TITLE
apt: Implement additional applet state management.

### PR DESCRIPTION
**Pull request by Steveice10**
Implements additional state management for applets to get closer to real APT behavior:
* APT now tracks the active applet slot, and uses it to in turn track which applet slot last launched a library applet.
* The last applet slot that launched a library is used for things like setting the correct source/destination applet ID in sent parameters.
* APT:Initialize will now check for an active applet slot, as real APT does, instead of its home menu workaround for sending the initial wake-up message.
* APT:GetLockHandle is now fully implemented, which can provide the correct applet attributes to a library applet caller based on what slot launched it.
* APT:CancelLibraryApplet has been implemented to allow applications to send a cancellation message to applets to abort launch.
* System applet commands have basic implementations, allowing at minimum the home menu to launch system applets.

There's also a secondary commit that adds clearing of a process's handle table when it closes, to prevent handle leaks. This resolves some hangs launching applets.

With these changes, the following issues (and possibly others) are fixed:
* When performing the initial system setup process, it would get stuck trying to load the EULA applet. This is because the settings app preloads the software keyboard beforehand. When launching EULA it would check IsRegistered for an existing library applet, which returned false because GetLockHandle was not correcting the applet attributes to point to the app library slot. Furthermore, it would be unable to close the applet without a CancelLibraryApplet implementation, and even with that the EULA applet would then hang launching due to dangling handles from the software keyboard process.
* Library applets now work from the home menu since the message from/to applet IDs are accurate, e.g. renaming a folder.
* System applets can be opened from the home menu. Some system applets have other issues that will need to be addressed in the future (e.g. Internet Browser)

Tested against a few relevant scenarios: booting home menu into the system setup, booting into the actual home menu, launching games from the home menu, launching games from the Citra game list, and using library applets within both the home menu and games.

- - -
This change is [Reviewable](https://reviewable.io/reviews/citra-emu/citra/6303)

### CODE REVIEW

**Comment by GPUCode** [comment]
Tested it alongside your other PRs and can confirm everything works. I was able to launch Face Raiders which I remember crashed before. Opening home menu settings though still crashes

**Comment by Steveice10** [comment]
> Tested it alongside your other PRs and can confirm everything works. I was able to launch Face Raiders which I remember crashed before. Opening home menu settings though still crashes

You may need to delete your config save game for the home menu settings in order to get it to reformat with the backlight config block added from one of my other PRs.

**Comment by GPUCode** [comment]
Tried nuking my citra folder to start from scratch for best results. Now though the home menu setup seems to be getting stuck when setting the profile (the sound stops and starts glitching too). The log fills with opengl errors so not sure if it's due to that or something else. Did you test this with vulkan only or opengl too?

**Comment by Steveice10** [comment]
I’ve only tested under Vulkan so far, it’s possible there are OpenGL issues that need to be resolved with the setup/menu.

**Comment by GPUCode** [comment]
I'll make a vulkan build then to make sure

**Comment by Steveice10** [comment]
Just to document, my steps for testing from a scratch Citra are:
* Add needed key files.
* Set region to USA (for myself, other regions should work too).
* Set renderer to Vulkan.
* Set 3D slider depth to maximum.
* Download All system files from NUS via the settings window.

When I do those steps it works as expected from start through the home menu.

**Comment by Steveice10** [comment]
Pushed in a small extra APT fix where extra bytes were being read into MediaType parameters from the command buffer, causing it to sometimes use garbage MediaType values when calling StartApplication.

**Comment by autofire372** [comment]
System Settings is crashing Citra when I get to internet settings with this.

**Comment by Steveice10** [comment]
> System Settings is crashing Citra when I get to internet settings with this.

I can't reproduce, internet settings opening fine on my end.

**Comment by autofire372** [comment]
System Settings is crashing Citra when I get to internet settings with this.

> 

I'll try with a fresh NAND in the morning.

**Comment by GPUCode** [comment]
> System Settings is crashing Citra when I get to internet settings with this.
> 
> > 
> 
> I'll try with a fresh NAND in the morning.

You might need to merge some other PRs, for example I need the key fix to prevent settings from crashing without aes_keys.txt

**Comment by autofire372** [comment]
I still can't proceed to internet settings, even with a fresh NAND. I started Citra from the command line, and when I hit next to go to internet settings, the terminal reported a stack-smash when Citra crashed. It did produce a core dump, if you want a look.

**Comment by Steveice10** [comment]
> I still can't proceed to internet settings, even with a fresh NAND. I started Citra from the command line, and when I hit next to go to internet settings, the terminal reported a stack-smash when Citra crashed. It's probably relevant that I'm running Linux (EndeavourOS).
> 
> Actually, here's what gdb had to say:

That’s helpful, thanks. Could you also share a Citra log with the filter set to “*:Debug” in the settings so I can get the debug logs?

**Comment by autofire372** [comment]
> That’s helpful, thanks. Could you also share a Citra log with the filter set to “*:Debug” in the settings so I can get the debug logs?

I actually just reproduced the crash on Windows 10, so it's not OS-specific. I'll get you that log shortly.

**Comment by Steveice10** [comment]
Managed to reproduce getting stuck at internet settings on a debug build due to some timing issues with how parameters were being sent. Rebased and pushed a new commit with a number of fixes for APT parameter passing accuracy.

* Sending parameters vs Cancelling and sending parameters vs Delayed parameter sending should now be accurate to what APT chooses to do for the currently implemented commands.
* HLE applets now follow the normal applet lifecycle a lot more closely, so calling Initialize/Enable/etc on launch and PrepareToCloseLibraryApplet/CloseLibraryApplet on exit. Everything is centered around sending the normal APT parameters to the HLE applet now, which has greatly reduced the number of special codepaths needed and fixed a bunch of APT state that wasn't tracked for HLE applets.
* Added some more debug logs to APT to help in future debugging.
* Fixed some existing incorrect IPC handling elsewhere that was blocking my debug builds from working properly.

Please let me know if this fixes your issues, and if not please share a new crash backtrace + log from the latest.

**Comment by autofire372** [comment]
> Managed to reproduce getting stuck at internet settings on a debug build due to some timing issues with how parameters were being sent. Rebased and pushed a new commit with a number of fixes for APT parameter passing accuracy.
> 
>     * Sending parameters vs Cancelling and sending parameters vs Delayed parameter sending should now be accurate to what APT chooses to do for the currently implemented commands.
> 
>     * HLE applets now follow the normal applet lifecycle a lot more closely, so calling Initialize/Enable/etc on launch and PrepareToCloseLibraryApplet/CloseLibraryApplet on exit. Everything is centered around sending the normal APT parameters to the HLE applet now, which has greatly reduced the number of special codepaths needed and fixed a bunch of APT state that wasn't tracked for HLE applets. This also let me implement CancelLibraryApplet properly for HLE applets which fixed HLE applets having the same hang as LLE applets did in the initial setup.
> 
>     * Added some more debug logs to APT to help in future debugging.
> 
>     * Fixed some existing incorrect IPC handling elsewhere that was blocking my debug builds from working properly.
> 
> 
> I've tested a few scenarios with this code like:
> 
>     * Initial setup with both HLE and LLE applets.
> 
>     * Home menu software keyboard with both HLE and LLE applets.
> 
>     * Mix of home menu library applets, system applets, and launching a game in one session.
> 
>     * Running games directly from the Citra game list with both HLE and LLE applets.
> 
>     * All of the above on debug (slower) builds to test against different timings.
> 
> 
> Please let me know if this fixes your issues, and if not please share a new crash backtrace + log from the latest.

I was able to complete setup, but was unable to proceed to the Home Menu afterward. Is this normal (i.e. not implemented yet)?

**Comment by Steveice10** [comment]
> > Managed to reproduce getting stuck at internet settings on a debug build due to some timing issues with how parameters were being sent. Rebased and pushed a new commit with a number of fixes for APT parameter passing accuracy.
> > ```
> > * Sending parameters vs Cancelling and sending parameters vs Delayed parameter sending should now be accurate to what APT chooses to do for the currently implemented commands.
> > 
> > * HLE applets now follow the normal applet lifecycle a lot more closely, so calling Initialize/Enable/etc on launch and PrepareToCloseLibraryApplet/CloseLibraryApplet on exit. Everything is centered around sending the normal APT parameters to the HLE applet now, which has greatly reduced the number of special codepaths needed and fixed a bunch of APT state that wasn't tracked for HLE applets. This also let me implement CancelLibraryApplet properly for HLE applets which fixed HLE applets having the same hang as LLE applets did in the initial setup.
> > 
> > * Added some more debug logs to APT to help in future debugging.
> > 
> > * Fixed some existing incorrect IPC handling elsewhere that was blocking my debug builds from working properly.
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > I've tested a few scenarios with this code like:
> > ```
> > * Initial setup with both HLE and LLE applets.
> > 
> > * Home menu software keyboard with both HLE and LLE applets.
> > 
> > * Mix of home menu library applets, system applets, and launching a game in one session.
> > 
> > * Running games directly from the Citra game list with both HLE and LLE applets.
> > 
> > * All of the above on debug (slower) builds to test against different timings.
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > Please let me know if this fixes your issues, and if not please share a new crash backtrace + log from the latest.
> 
> I was able to complete setup, but was unable to proceed to the Home Menu afterward. Is this normal (i.e. not implemented yet)?

Yes, the home button is not implemented yet so you just have to re-launch the home menu when the system setup is done.

**Comment by FearlessTobi** [comment]
This is causing regressions when using applets in Tomodachi Life.
(for more info see the #dev channel on discord)

**Comment by Steveice10** [comment]
Did a general cleanup of current APT code and it seems to have resolved the issue, at least on the Android device where I could reproduce it.

**Comment by Steveice10** [comment]
Latest commit fixes the Game Notes system applet when it has no save data.

File: src/core/hle/service/frd/frd.cpp [hunk 1]
```diff
@@ -94,7 +94,7 @@ void Module::Interface::GetMyFriendKey(Kernel::HLERequestContext& ctx) {
 
 void Module::Interface::GetMyScreenName(Kernel::HLERequestContext& ctx) {
     IPC::RequestParser rp(ctx, 0x9, 0, 0);
-    IPC::RequestBuilder rb = rp.MakeBuilder(7, 0);
+    IPC::RequestBuilder rb = rp.MakeBuilder(6, 0);
```
Comment (PabloMK7) [inline src/core/hle/service/frd/frd.cpp:97]: This change is wrong, should be 7 not 6.
Screen name is 11 UTF-16 characters, so 22 bytes or 6 words, plus the return code.
Also verified in the friends module disassembly.

Comment (Steveice10) [reply src/core/hle/service/frd/frd.cpp]: Fixed, thanks.

File: src/core/hle/service/apt/apt.cpp [hunk 2]
```diff
@@ -420,42 +424,40 @@ void Module::APTInterface::GlanceParameter(Kernel::HLERequestContext& ctx) {
     LOG_DEBUG(Service_APT, "called app_id={:#010X}, buffer_size={:#010X}", app_id, buffer_size);
 
     auto next_parameter = apt->applet_manager->GlanceParameter(app_id);
-
     if (next_parameter.Failed()) {
         IPC::RequestBuilder rb = rp.MakeBuilder(1, 0);
         rb.Push(next_parameter.Code());
-        return;
+    } else {
```
Comment (FearlessTobi) [inline src/core/hle/service/apt/apt.cpp:430]: i don't see why this needs to be changed. dito below

**Review by FearlessTobi — COMMENTED** [review]
Apart from this nit, the code looks good

File: src/core/hle/service/apt/apt.cpp (hunk 2 above)
Comment (Steveice10) [reply src/core/hle/service/apt/apt.cpp:430]: Matching style with some other IPC functions that use the same pattern in this file. I can change it back if you would prefer.

Comment (FearlessTobi) [reply src/core/hle/service/apt/apt.cpp:430]: honestly doesn't matter too much.
you can leave it as is

**Pull request merged**